Design.md

```diff
@@ -0,0 +1,177 @@
+
+## Resource limits and isolation
+
+The job-runner uses [cgroups V2](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html) for resource limits. There are many tweakable options for limiting resource use, but the following minimal set has been selected to demonstrate how they work. In particular, the following cgroup parameters will be set:
+* CPU:
+    * `cpu.max` quota and period
+* Memory:
+    * `memory.max` (max amount of memory to be used, hard limit)
+    * `memory.high` (set to 75% of `memory.max`)
+* Disk IO: (Device list readable from `/proc/partitions`)
+    * `io.max.{rbps,wbps,riops,wiops}`
+
+In addition to limiting resource use with cgroups, job-runner also isolates jobs from each other using namespaces. A PID namespace is set up to make sure the job cannot kill processes not spawned by it, and to make sure all child processes are terminated together with the actual job. A mount namespace is used to limit process to a subset of the file system, together with [`pivot_root(2)`](https://linux.die.net/man/2/pivot_root) (see [Understanding Containerization By Recreating Docker](https://itnext.io/linux-container-from-scratch-339c3ba0411d), search for pivot_root). Finally, a network namespace is created to limit network access of the jobs. It must be used together with [Virtual ethernet (VETH)](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#veth) and bridge interfaces if jobs should be allowed to communicate between each other. This also means that internet access must be granted separately.
+
+## Communication and auth
+
+Clients communicate via gRPC with a simple protocol: One service with four endpoints. The communication is secured [RusTLS](https://github.com/ctz/rustls) which properly [audited](https://github.com/ctz/rustls/blob/master/audit/TLS-01-report.pdf) and by design supports only modern, secure cipher suites. Only TLS 1.2/1.3 are used, and authentication is done using ECDSA, Ed25519 or RSA. RusTLS will remove support for cipher suites deemed insecure, and simply keeping the version up to date should be sufficient in the future as well. [Tonic](https://github.com/hyperium/tonic) is used to serve gRPC and almost automatically handles protocol buffers, encryption and related concerns.
+
+Authentication is implemented with mTLS. Server and client have different CA roots, which they are expected to exchange in a secure way. (Scenario-wise: the client CA is operated by the same organization that hosts job-runner). The server identifies each client with it's CommonName (CN) field of the certificate. The client CA only issues certificates with CN values to developers that are allowed to access the API. All running jobs are bound to the CN of the client calling `Start`, and only a client with that certificate is allowed to execute operations for that job.
```

Comment:
It sounds like what you're proposing is any authenticated user can create a job but you can only call stop/status/logs on your own job? Is this correct?

Reply:
Correct. I clarified the document as well.